### PR TITLE
Fix #210: create sirepo_sandbox

### LINKED ISSUE #210: sirepo_sandbox component
A sirepo_sandbox is almost like mpi_worker: a docker container that runs sshd for a single user. The purpose is to allow a user to develop sirepo, using ssh to copy files.

Unlike mpi_worker, there need to be multiple instances on the node. This is similar to how bop starts multiple apps. However, these users need different ports to ssh into. 

The network should be isolated so only ssh will work, and everything else is tunneled into and out of the container via ssh. This would allow a user to  connect to the development server on the node from a local browser. We will need to use docker proxy for this.

Files have to persist so there will be read-write volumes mounted, specifically ~/.ssh, ~/src/radiasoft, and ~/jupyter. The latter is used to persist ~/jupyter/pre_jupyter_bashrc and ~/jupyter/bashrc. That's a convenient hook for this problem. No need to invent something new for this problem.

Already encrypted private keys will need to be distributed to end users outside the rsconf configured nodes. This is different than mpi_worker which uses non-encrypted keys and the distribution happens to external devices (laptops). The main point is the key is encrypted by sirepo_sandbox component with a passphrase that is configured and transferred to users along with the keys. The protocol can be manual as it is all happening inside the organization.

### PR DESCRIPTION
**Pull request by e-carlin**
@robnagler this creates the sirepo_sandbox component. I have it working (can ssh in using identity file). But, there are 2 TODO's below that I could use your thoughts on (1. how to get /srv/ inside of the component, 2. how to mount ~/src as a volume).  

Example db/000.yml config
```yaml
sirepo_sandbox:
  docker_image: radiasoft/beamsim-jupyter
  guest_d: /home/vagrant/
  # TODO(e-carlin): this should come from rsconf (see comment in sirepo_sandbox.py)
  host_root_d: /srv/
  users:
    e-carlin: 9999
    nagler: 8888
  volumes:
    - jupyter
    # TODO(e-carlin): doesn't work (problem with home-env not existing)
    # - src
```

Example db/secret/000.yml config
```yaml
sirepo_sandbox:
  user_passwords:
    nagler: pass-nagler
      e-carlin: pass-e-carlin

### CODE REVIEW

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [inline rsconf/component/sirepo_sandbox.py:104]: You said:

> Already encrypted private keys will need to be distributed to end users outside the rsconf configured nodes. This is different than mpi_worker which uses non-encrypted keys and the distribution happens to external devices (laptops). The main point is the key is encrypted by sirepo_sandbox component with a passphrase that is configured and transferred to users along with the keys. The protocol can be manual as it is all happening inside the organization.

I wasn't 100% sure what you meant but is this it? We can manually create passwords and populate them in db/secret/000.yml, run `rsconf build`, then manually distribute the password/key to users?

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:1]: I think this should call this devbox. sirepo_sandbox is tool long. The instances would be devbox_ecarlin, devops_nagler, etc.

File: rsconf/package_data/sirepo_sandbox/sshd_config.jinja
Comment (robnagler) [inline rsconf/package_data/sirepo_sandbox/sshd_config.jinja:5]: Not needed.

File: rsconf/package_data/sirepo_sandbox/sshd_config.jinja
Comment (robnagler) [inline rsconf/package_data/sirepo_sandbox/sshd_config.jinja:6]: mpi_worker again, we are trying to control things so mpi works. for "normal" logins, this is fine.

File: rsconf/package_data/sirepo_sandbox/sshd_config.jinja
Comment (robnagler) [inline rsconf/package_data/sirepo_sandbox/sshd_config.jinja:8]: We do want this, because it contains a disclaimer. mpi_worker is different. We want users to think the are logging into a "real" machine.

File: rsconf/package_data/sirepo_sandbox/sshd_config.jinja
Comment (robnagler) [inline rsconf/package_data/sirepo_sandbox/sshd_config.jinja:10]: I don't think we need this. People will be logging in from real computers on the internet.

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:79]: I think self.name will be devbox_<name>. I think we have to cascade a different key (user_name) for this.

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:37]: fail fast: `z[x + "_d']`

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:118]: I don't think we want this as a dot file. This could be called sshd and then mounted in the guest that way.

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [reply rsconf/component/sirepo_sandbox.py:105]: I think we can generate passwords and it will be simpler than configuring them. They do not need to be distributed, just dumped to json. Just like rsconf does, for example. These passwords will be changed by the user once the private key is on their machine.

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:87]: I don't think this needs to be on the channel, it can be just on the host. There's no channel relationship that needs to be maintained. Fix secret_path to take a parameter "directory=True" in this case that would do a mkdir_parent instead of a mkdir_parent_only. This is probably used in other places.

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:85]: share all this with mpi_worker in `component.__init__`

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:23]: pass in user in kwargs to T

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:47]: rather ": user instances do all the installs"

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:55]: We also want db/src which is rw and owned by vagrant.

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:53]: This should be read-only too.

File: rsconf/component/sirepo_sandbox.py
Comment (robnagler) [inline rsconf/component/sirepo_sandbox.py:73]: You need to install only the private key for the host and the public key for the user. Not both. Unlike mpi_worker, the public key for the host and the private key for ssh will reside on the user's laptop (not `~/jupyter`).

**Review by robnagler — CHANGES_REQUESTED** [review]
We need to talk about user volumes.

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py]: 566ec33

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: 566ec33

File: rsconf/package_data/sirepo_sandbox/sshd_config.jinja
Comment (e-carlin) [reply rsconf/package_data/sirepo_sandbox/sshd_config.jinja]: 566ec33

File: rsconf/package_data/sirepo_sandbox/sshd_config.jinja
Comment (e-carlin) [reply rsconf/package_data/sirepo_sandbox/sshd_config.jinja]: 566ec33

File: rsconf/package_data/sirepo_sandbox/sshd_config.jinja
Comment (e-carlin) [reply rsconf/package_data/sirepo_sandbox/sshd_config.jinja]: 566ec33

File: rsconf/package_data/sirepo_sandbox/sshd_config.jinja
Comment (e-carlin) [reply rsconf/package_data/sirepo_sandbox/sshd_config.jinja]: 566ec33

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: 566ec33

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: aeab41b

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: aeab41b

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: aeab41b

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: aeab41b

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: af3b1f8

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: 147cc16

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: 1f49f6a

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: 1f49f6a

**Comment by robnagler** [comment]
@e-carlin what's the status? I think we need to get this moving into production so we can start coding.

**Comment by e-carlin** [comment]
I need to address the last couple of comments. I'll fix and request another review by EOD

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: 8d1bb90

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: 8d68e84

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: They are included through "volumes" configuration (although I added rw 2eccc4d). There is a problem with src. The container exits immediately upon start with no message. My guess is it is because it is missing home-env and something calls some function from it on start. 

I need to fix https://github.com/radiasoft/rsconf/issues/224 and then can test more.

**Comment by e-carlin** [comment]
@robnagler ready for another review. All problems solved except https://github.com/radiasoft/rsconf/pull/217#discussion_r936085938 which we should talk about.

File: rsconf/component/sirepo_sandbox.py
Comment (e-carlin) [reply rsconf/component/sirepo_sandbox.py]: [d2604ae](https://github.com/radiasoft/rsconf/pull/217/commits/d2604ae449df7786c57f4ee0ce59e23414c19ed3)

File: rsconf/component/__init__.py
Comment (robnagler) [inline rsconf/component/__init__.py:340]: I think `b.new(ext="pub")` is how you are supposed to do this in py.path.

File: rsconf/component/__init__.py
Comment (robnagler) [inline rsconf/component/__init__.py:332]: Seems like this should be public.

File: rsconf/component/__init__.py
Comment (robnagler) [inline rsconf/component/__init__.py:333]: Default to ""?

File: rsconf/component/__init__.py
Comment (robnagler) [inline rsconf/component/__init__.py:335]: could be imported up top

File: rsconf/component/devbox.py
Comment (robnagler) [inline rsconf/component/devbox.py:102]: This will change the password every time. You want to do this first so you can get the password for the user or create it. See component.rsconf.host_init.

File: rsconf/systemd.py
Comment (robnagler) [inline rsconf/systemd.py:76]: Use a temporary or use `if not run_d: run_d = ...`.  I don't like assigning run_d if already assigned.

File: rsconf/component/devbox.py
Comment (robnagler) [inline rsconf/component/devbox.py:33]: seems like compt.name could be devbox_user_name and set that on create. Then everything will be consistent.

**Review by robnagler — CHANGES_REQUESTED** [review]
Would be good to add some cases to tests. I think we might want to do something with random_string to get reliable output, but you could skip that for now

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py:332]: I get tripped up in Python with classes and what should be public but I believe since only instances of the class call this method (using self or super) private is correct. Agree?

```python
rsconf (210-sirepo-sandbox)$ gpy _gen_host_and_identity_ssh_keys
./rsconf/component/__init__.py:332:    def _gen_host_and_identity_ssh_keys(
./rsconf/component/devbox.py:34:        z.secrets = self._gen_host_and_identity_ssh_keys(jc)
./rsconf/component/devbox.py:85:    def _gen_host_and_identity_ssh_keys(self, jc):
./rsconf/component/devbox.py:89:        res = super()._gen_host_and_identity_ssh_keys(
./rsconf/component/mpi_worker.py:26:        z.secrets = self._gen_host_and_identity_ssh_keys(jc, z)
./rsconf/component/mpi_worker.py:95:    def _gen_host_and_identity_ssh_keys(self, jc, z):
./rsconf/component/mpi_worker.py:96:        return super()._gen_host_and_identity_ssh_keys(
./rsconf/component/mpi_worker.py:130:            s = self._gen_host_and_identity_ssh_keys(jc, z)
```

File: rsconf/component/devbox.py
Comment (e-carlin) [reply rsconf/component/devbox.py]: 0b2eeca

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py]: 0b2eeca

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py]: 0b2eeca

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py]: 0b2eeca

File: rsconf/component/devbox.py
Comment (e-carlin) [reply rsconf/component/devbox.py]: 0b2eeca

File: rsconf/component/devbox.py
Comment (e-carlin) [reply rsconf/component/devbox.py]: fixing the compt.name removed the need for supplying run_d 5d2406d

Comment (e-carlin) [reply rsconf/component/devbox.py]: Whoops, previous comment was on the wrong comment of yours. 

Thanks for seeing compt.name. Cleaned things up 5d2406d

File: rsconf/systemd.py
Comment (e-carlin) [reply rsconf/systemd.py]: fixing the compt.name removed the need for supplying run_d https://github.com/radiasoft/rsconf/commit/5d2406d37cba0a5b455f327bc6df141c05c6010a

File: rsconf/component/__init__.py
Comment (robnagler) [reply rsconf/component/__init__.py:329]: It's inconsistent with the other methods in component.T. I understand the principle of most private. The only guidance in PEP8 is: 

> Use one leading underscore only for non-public methods and instance variables.

and

> Another category of attributes are those that are part of the “subclass API” (often called “protected” in other languages). Some classes are designed to be inherited from, either to extend or modify aspects of the class’s behavior. When designing such a class, take care to make explicit decisions about which attributes are public, which are part of the subclass API, and which are truly only to be used by your base class.

Underscores do not seem to come into play here other than reading between the lines.

I think there is no such thing as "protected" in Python. In BOP we labeled methods with an "internal_" prefix to mean protected, but Perl adopts no such convention and there is no name mangling like in Python for double underscore.

In this particular case, I vote for consistency for the other functions. You certainly can add this as a /discuss issue.

File: rsconf/component/devbox.py
Comment (robnagler) [inline rsconf/component/devbox.py:27]: nice :)

File: rsconf/component/devbox.py
Comment (robnagler) [inline rsconf/component/devbox.py:91]: I liked your module_name concept. Maybe this could be part of the creation of the instance in the base class so that a component has a name and a module_name, and then overriden above when creating the child instances.

File: rsconf/package_data/rsconf/rsconf.sh
Comment (robnagler) [inline rsconf/package_data/rsconf/rsconf.sh:47]: Why the need for the second check? If it is a directory, it should not be cloned.

File: rsconf/package_data/rsconf/rsconf.sh
Comment (robnagler) [inline rsconf/package_data/rsconf/rsconf.sh:5]: Thanks.

File: rsconf/package_data/rsconf/rsconf.sh
Comment (robnagler) [inline rsconf/package_data/rsconf/rsconf.sh:46]: user => owner (and implicitly group). 

Perhaps we need an install umask for this case. 

Also set the umask in a subshell and run git clone  In this case we can hardwire 027 as the umask for now until we need something different.

Note the weirdness about [git clone and umask](https://stackoverflow.com/a/11335009).

**Review by robnagler — CHANGES_REQUESTED** [review]

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py]: Make sense. Thanks for the details. 367ae5c

File: rsconf/package_data/rsconf/rsconf.sh
Comment (e-carlin) [reply rsconf/package_data/rsconf/rsconf.sh:47]: We need to be able to handle the fact that `devbox_<user>.sh` is going to create the full directory structure. Ex
```
rsconf_install_access '700' 'vagrant' 'vagrant'
rsconf_install_directory '/srv/devbox_xxx/jupyter'
rsconf_install_directory '/srv/devbox_xxx/src'
rsconf_install_directory '/srv/devbox_xxx/src/biviosoftware'
rsconf_install_directory '/srv/devbox_xxx/src/biviosoftware/home-env'
rsconf_clone_repo https://github.com/biviosoftware/home-env.git /srv/devbox_xxx/src/biviosoftware/home-env vagrant
```

File: rsconf/package_data/rsconf/rsconf.sh
Comment (e-carlin) [reply rsconf/package_data/rsconf/rsconf.sh]: 9ffe221

Looks like that weirdness was fixed in [1.7.11.3](https://github.com/git/git/commit/45d4fdc2dc74c657a9c2e95bf04aed539fdcb0a4). We have 1.8.3.1 on centos7.

File: rsconf/package_data/rsconf/rsconf.sh
Comment (robnagler) [reply rsconf/package_data/rsconf/rsconf.sh:47]: OK.

File: rsconf/component/devbox.py
Comment (e-carlin) [reply rsconf/component/devbox.py]: I'm not sure I understand what you're looking for. Is this it 9d156a0 ?

File: rsconf/component/mpi_worker.py
Comment (robnagler) [inline rsconf/component/mpi_worker.py:18]: z.user could be assigned self.user_name, or devbox could put user in z. The goal is to share as much code with gen_host_and_identity_ssh_keys. I don't think this needs to be a subclassed method.

**Review by robnagler — CHANGES_REQUESTED** [review]
I gave a bunch of comments on the commit, which means you are going to have to find them. I edited one of the comments.

File: rsconf/component/mpi_worker.py
Comment (e-carlin) [reply rsconf/component/mpi_worker.py]: How does this look ccdb931 ?

File: rsconf/component/__init__.py
Comment (robnagler) [inline rsconf/component/__init__.py:88]: I think just "encrypt_identity=False" would be better, and have the password_filename be `f"{self.module_name}_ssh_passphrase.json"`. It's not a password, and it isn't "devbox_auth". It's a passphrase for the ssh identity key, but that's long so ssh_passphrase is clear, I think.

File: rsconf/component/__init__.py
Comment (robnagler) [inline rsconf/component/__init__.py:90]: Too complicated and unnecessary work. You can write the passphrase before the generate happens. Logic should be:

```py
for f:
    if not f.exists():
        process_call(
        ...
       _pass() if encrypt_identity and f == res.identity_f else "",
```
_pass would be simpler:

```py    
def _pass():
     s = db.secret_path(j2_ctx, f"{self.module_name}_ssh_passphrase.json", visibility=visibility)
     o = pkjson.load_any(s) if s.exists() else PKDict()
     p = o.get(self.user_name)
     if p is None:
         o[self.user_name] = p = db.random_string()
         pkjson.dump_pretty(o, filename=s)
     return p
```

File: rsconf/component/__init__.py
Comment (robnagler) [inline rsconf/component/__init__.py:111]: Could be simplified:

```py
for x in "identity", "host_key":
    # possibly make this a function
    y = tuple((f"{x}{s}"_f", b.join(x).new(ext=s[1:]) for s in "", "_pub")
    res.update(y)
    if y[0][1].exists:
        continue
```

File: rsconf/component/__init__.py
Comment (robnagler) [inline rsconf/component/__init__.py:213]: I think there should always be a module_name. For backward compatibility, you will need to only do this if "/" isn't already in name iwc the caller already added module_name (typical case). This is a highly constrained problem, because we never have a "name" without a "/" (you can assert this), that is, all resources are at least two levels deep. Therefore I think this trick will be sufficient to fix all cases.

**Review by robnagler — CHANGES_REQUESTED** [review]

File: rsconf/component/devbox.py
Comment (e-carlin) [reply rsconf/component/devbox.py]: 5cf71b2

File: rsconf/component/devbox.py
Comment (e-carlin) [reply rsconf/component/devbox.py]: 5cf71b2

File: rsconf/component/devbox.py
Comment (e-carlin) [reply rsconf/component/devbox.py]: 5cf71b2

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py]: 5cf71b2

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py]: Much better, thanks 5cf71b2

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py]: > you will need to only do this if "/" isn't already in name

I'm not sure what you mean by "this"?

I think just using self.module_name works 749c4cb. We default module_name to name. That I believe satisfies our backwards compat needs. In the devbox case module_name is devbox (not devbox_<username> which is what self.name is) so it works in that case as well.

File: rsconf/component/__init__.py
Comment (e-carlin) [reply rsconf/component/__init__.py]: 5cf71b2

**Review by robnagler — APPROVED** [review]
Not the day to push to production so I'll let you merge on Monday.

**Comment by robnagler** [comment]
Tests are failing so please fix before merge.

**Pull request merged**
